### ISSUE TITLE
omprog not properly closing child process (similar to #2599, in part) 

### ISSUE THREAD

**Issue by systemdlete** [rsyslog/rsyslog#5300]
### Expected behavior
Child process called by omprog action should receive EOF when output from parent (rsyslogd itself) ends.

### Actual behavior
Child process hangs on read on stdin.  Note that the receiving program does get input, but it does not receive a terminating EOF.

### Steps to reproduce the behavior
Pass any multi-line output from rsyslogd to a bash script:
action(type="omprog" binary="/bin/bash /path/to/my/bash-script")

### Environment
- rsyslog version: 8.2302.0
- platform: devuan daedalus

Commit #2629 was supposed to fix this, but the behavior described persists in this version (8.2302.0).  This bug is similar to the one reported in #2599, but in this case, the flag "signalOnClose" is not passed to omprog.

**Comment by davidelang** [comment]
how is rsyslog supposed to know that there will never be another log message to 
send to the omprog process? it doesn't start a new instance of omprog for every 
log message, it starts one and keeps sending new logs to it until rsyslog exits 
or has a restart-like event

I may be missing something here, but it sounds like it's working as designed. 
What am I missing?

David Lang

**Comment by systemdlete** [comment]
Apologies if I misunderstood the intention of omprog.   The documentation does imply that the "binary" (i.e., the program to be executed by omprog) will be restarted if it terminates, but it might be better if it specified something like the binary is expected to be a server daemon of some kind, a persistent program that runs "forever."   For people who are accustomed to calls like system(3), and other tools which write to, say, a pipe or socket, omprog at first SOUNDS like something similar, but it is not; for me (and probably others like myself), only a very close and maybe repeated reading of the rather terse documentation followed by a long period of contemplation helps to clarify the difference.

I had "resolved" the problem by including a fake EOF in the sent data, but now I see your real intention.   So I will re-write my called program ("binary") to run as a long-runner, non-terminating sort of server program.   Launching it once and running it persistently does, indeed, make more sense than firing off the same "binary" repeatedly.  It just wasn't obvious to me at the get-go that this was the intention.   Again, sorry for the noise.

**Comment by davidelang** [comment]
Yes, it will restart the daemon if it dies, but that's a backup mechanism, not 
the primary way to differenitate logs from each other.

Do you have suggestions on how to make the docs clearer

too many of us who work on the docs are too close to the code, and don't always 
see where things can be read to imply something different than how we 'know' the 
system works. Outside suggestions (and even patches to the docs) are extremely 
welcome.

David Lang

**Comment by systemdlete** [comment]
As I suggested above, the documentation might be explicit such as "the binary is expected to be a server daemon of some kind, a persistent program that runs forever."   (You might want to reword that to be more precise, idk.)

BTW, does that "binary" have to be a true *nix binary executable, or does omprog support hash-bang (#!) such as in scripts?  Currently, my "binary" is /bin/bash and its argument in the call from omprog is the complete path to the bash script I want to run.

**Comment by davidelang** [comment]
On Sat, 23 Dec 2023, systemdlete wrote:

> As I suggested above, the documentation might be explicit such as "the binary is expected to be a server daemon of some kind, a persistent program that runs forever."   (You might want to reword that to be more precise, idk.)
>
> BTW, does that "binary" have to be a true *nix binary executable, or does omprog support hash-bang (#!) such as in scripts?  Currently, my "binary" is /bin/bash and its argument in the call from omprog is the complete path to the bash script I want to run.

it supports #! scripts, that's a common thing to do.

David Lang

**Comment by rgerhards** [comment]
Apologies for the late reply. This may no longer be relevant for the original report, but I am adding the outcome here for people who arrive via search while we work through the backlog. We do not close old issues lightly or just because of age.

The discussion resolved this as expected `omprog` behavior rather than a bug. `omprog` starts the configured program and keeps it running so messages can be streamed to it over time; it does not start a new process for each message and then close stdin after each write. EOF is expected when rsyslog exits, restarts, or otherwise closes the action process, not as a per-message delimiter. The reporter confirmed this understanding and explicitly said the issue could be closed.

A useful note from the thread: the `binary` can be a normal executable or a script using a `#!` interpreter line, and a long-running program is the intended model. I am closing this as answered. If the current documentation should be clarified further, that is best handled as a focused doc issue or PR with suggested wording.
